File: src/parser.py
from src.token_types_and_keywords import *
from src.nodes import *
from src.token import Token
from src.errors import InvalidSyntaxError


class ParseResult:
    def __init__(self):
        self.error = None
        self.node = None
        self.last_registered_advance_count = 0
        self.advance_count = 0
        self.to_reverse_count = 0

    def registerAdvancement(self):
        self.last_registered_advance_count = 1
        self.advance_count += 1

    def register(self, res):
        self.last_registered_advance_count = res.advance_count
        self.advance_count += res.advance_count

        if res.error:
            self.error = res.error

        return res.node

    def tryRegister(self, res):
        if res.error:
            self.to_reverse_count = res.advance_count

            return None

        return self.register(res)

    def success(self, node):
        self.node = node
        return self

    def failure(self, error):
        if not self.error or self.last_registered_advance_count == 0:
            self.error = error
        return self


class Parser:
    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.token_index = -1

        self.advance()

    def advance(self):
        self.token_index += 1
        self.updateCurrentToken()
        return self.current_token

    def reverse(self, amount=1):
        self.token_index -= amount
        self.updateCurrentToken()
        return self.current_token

    def updateCurrentToken(self):
        if self.token_index >= 0 and self.token_index < len(self.tokens):
            self.current_token = self.tokens[self.token_index]

    def parse(self):
        result = self.statements()

        if not result.error and self.current_token.type is not TT_EOF:
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected operators "+", "-", "*", "/" or "endbody"'
            ))

        return result

    def listExpr(self):
        result = ParseResult()
        element_nodes = []
        pos_start = self.current_token.pos_start.copy()

        if self.current_token.type != TT_LSQUARE:
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "["'
            ))

        result.registerAdvancement()
        self.advance()

        if self.current_token.type == TT_RSQUARE:
            result.registerAdvancement()
            self.advance()

        else:
            element_nodes.append(result.register(self.expr()))

            if result.error:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected "]", "object", "if", "walk", "while", "func", int, float, identifier, "+", "-", "(", "[" or "oppositeof"'
                ))

            while self.current_token.type == TT_COMMA:
                result.registerAdvancement()
                self.advance()

                element_nodes.append(result.register(self.expr()))

                if result.error:
                    return result

            if self.current_token.type != TT_RSQUARE:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected "," or "]"'
                ))

            result.registerAdvancement()
            self.advance()

        return result.success(ListNode(
            element_nodes,
            pos_start,
            self.current_token.pos_end.copy()
        ))

    def ifExpr(self):
        result = ParseResult()
        all_cases = result.register(self.ifExprCases('if'))

        if result.error:
            return result

        cases, else_case = all_cases
        return result.success(IfNode(cases, else_case))

    def ifExpr_b(self):
        return self.ifExprCases('alsoif')

    def ifExpr_c(self):
        result = ParseResult()
        else_case = None

        if self.current_token.matches(TT_KEYWORD, 'otherwise'):
            result.registerAdvancement()
            self.advance()

            if self.current_token.type == TT_NEWLINE:
                result.registerAdvancement()
                self.advance()

                statements = result.register(self.statements())

                if result.error:
                    return result

                else_case = (statements, True)

                if self.current_token.matches(TT_KEYWORD, 'endbody'):
                    result.registerAdvancement()
                    self.advance()

                else:
                    return result.failure(InvalidSyntaxError(
                        self.current_token.pos_start, self.current_token.pos_end,
                        'Expected "endbody"'
                    ))

            else:
                expr = result.register(self.statement())

                if result.error:
                    return result

                else_case = (expr, False)

        return result.success(else_case)

    def ifExpr_b_or_c(self):
        result = ParseResult()
        cases, else_case = [], None

        if self.current_token.matches(TT_KEYWORD, 'alsoif'):
            all_cases = result.register(self.ifExpr_b())

            if result.error:
                return result

            cases, else_case = all_cases

        else:
            else_case = result.register(self.ifExpr_c())

            if result.error:
                return result

        return result.success((cases, else_case))

    def ifExprCases(self, case_keyword):
        result = ParseResult()
        cases = []
        else_case = None

        if not self.current_token.matches(TT_KEYWORD, case_keyword):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                f'Expected "{case_keyword}"'
            ))

        result.registerAdvancement()
        self.advance()

        condition = result.register(self.statement())

        if result.error:
            return result

        if not self.current_token.matches(TT_KEYWORD, 'then'):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "then"'
            ))

        result.registerAdvancement()
        self.advance()

        if self.current_token.type == TT_NEWLINE:
            result.registerAdvancement()
            self.advance()

            statements = result.register(self.statements())

            if result.error:
                return result

            cases.append((condition, statements, True))

            if self.current_token.matches(TT_KEYWORD, 'endbody'):
                result.registerAdvancement()
                self.advance()

            else:
                all_cases = result.register(self.ifExpr_b_or_c())

                if result.error:
                    return result

                new_cases, else_case = all_cases
                cases.extend(new_cases)

        else:
            expr = result.register(self.expr())

            if result.error:
                return result

            cases.append((condition, expr, False))

            all_cases = result.register(self.ifExpr_b_or_c())

            if result.error:
                return result

            new_cases, else_case = all_cases
            cases.extend(new_cases)

        return result.success((cases, else_case))

    def forExpr(self):
        result = ParseResult()

        if not self.current_token.matches(TT_KEYWORD, 'walk'):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "walk"'
            ))

        result.registerAdvancement()
        self.advance()

        if self.current_token.type != TT_IDENTIFIER:
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected identifier'
            ))

        var_name = self.current_token
        result.registerAdvancement()
        self.advance()

        if self.current_token.type != TT_EQ:
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "="'
            ))

        result.registerAdvancement()
        self.advance()

        start_value = result.register(self.expr())

        if result.error:
            return result

        if not self.current_token.matches(TT_KEYWORD, 'through'):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "through"'
            ))

        result.registerAdvancement()
        self.advance()

        end_value = result.register(self.expr())

        if result.error:
            return result

        if self.current_token.matches(TT_KEYWORD, 'step'):
            result.registerAdvancement()
            self.advance()

            step_value = result.register(self.expr())

            if result.error:
                return result

        else:
            step_value = None

        if not self.current_token.matches(TT_KEYWORD, 'then'):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "then"'
            ))

        result.registerAdvancement()
        self.advance()

        if self.current_token.type == TT_NEWLINE:
            result.registerAdvancement()
            self.advance()

            body = result.register(self.statements())

            if result.error:
                return result

            if not self.current_token.matches(TT_KEYWORD, 'endbody'):
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.advance().pos_end,
                    'Expected "endbody"'
                ))

            result.registerAdvancement()
            self.advance()

            return result.success(ForNode(var_name, start_value, end_value, step_value, body, True))

        body = result.register(self.statement())

        if result.error:
            return result

        return result.success(ForNode(var_name, start_value, end_value, step_value, body, False))

    def whileExpr(self):
        result = ParseResult()

        if not self.current_token.matches(TT_KEYWORD, 'while'):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "while"'
            ))

        result.registerAdvancement()
        self.advance()

        condition = result.register(self.expr())

        if result.error:
            return result

        if not self.current_token.matches(TT_KEYWORD, 'then'):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "then"'
            ))

        result.registerAdvancement()
        self.advance()

        if self.current_token.type == TT_NEWLINE:
            result.registerAdvancement()
            self.advance()

            body = result.register(self.statements())

            if result.error:
                return result

            if not self.current_token.matches(TT_KEYWORD, 'endbody'):
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected "endbody"'
                ))

            result.registerAdvancement()
            self.advance()

            return result.success(WhileNode(condition, body, True))

        body = result.register(self.statement())

        if result.error:
            return result

        return result.success(WhileNode(condition, body, False))

    def arithmeticExpr(self):
        return self.binaryOperator(self.term, (TT_PLUS, TT_MINUS))

    def comparisonExpr(self):
        result = ParseResult()

        if self.current_token.matches(TT_KEYWORD, 'oppositeof'):
            op_token = self.current_token
            result.registerAdvancement()
            self.advance()

            node = result.register(self.comparisonExpr())

            if result.error:
                return result

            return result.success(UnaryOperatorNode(op_token, node))

        node = result.register(self.binaryOperator(self.arithmeticExpr, (TT_EE, TT_NE, TT_LT, TT_GT, TT_LTE, TT_GTE)))

        if result.error:
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected int, float, identifier, "+", "-", "(", "[" or "oppositeof"'

            ))

        return result.success(node)

    def expr(self):
        result = ParseResult()

        if self.current_token.matches(TT_KEYWORD, 'object'):
            result.registerAdvancement()
            self.advance()

            if self.current_token.type != TT_IDENTIFIER:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected identifier'
                ))

            var_name = self.current_token
            result.registerAdvancement()
            self.advance()

            if self.current_token.type != TT_EQ:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected "="'
                ))

            result.registerAdvancement()
            self.advance()
            expr = result.register(self.expr())

            if result.error:
                return result

            return result.success(VariableAssignNode(var_name, expr))

        node = result.register(self.binaryOperator(self.comparisonExpr, ((TT_KEYWORD, 'and'), (TT_KEYWORD, 'or'))))

        if result.error:
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "object", "if", "walk", "while", "func", int, float, identifier, "+", "-", "(", "[" or "oppositeof"'
            ))

        return result.success(node)

    def statement(self):
        result = ParseResult()
        pos_start = self.current_token.pos_start.copy()

        if self.current_token.matches(TT_KEYWORD, 'give'):
            result.registerAdvancement()
            self.advance()

            expr = result.tryRegister(self.expr())

            if not expr:
                self.reverse(result.to_reverse_count)

            return result.success(ReturnNode(expr, pos_start=pos_start, pos_end=self.current_token.pos_start.copy()))

        if self.current_token.matches(TT_KEYWORD, 'next'):
            result.registerAdvancement()
            self.advance()

            return result.success(ContinueNode(pos_start=pos_start, pos_end=self.current_token.pos_start.copy()))

        if self.current_token.matches(TT_KEYWORD, 'leave'):
            result.registerAdvancement()
            self.advance()

            return result.success(BreakNode(pos_start=pos_start, pos_end=self.current_token.pos_start.copy()))

        expr = result.register(self.expr())

        if result.error:
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "give", "next", "leave", "object", "if", "walk", "while", "func", int, float, identifier, '
                '"+", "-", "(", "[" or "oppositeof"'
            ))

        return result.success(expr)

    def statements(self):
        result = ParseResult()
        statements = []
        pos_start = self.current_token.pos_start.copy()

        while self.current_token.type == TT_NEWLINE:
            result.registerAdvancement()
            self.advance()

        statement = result.register(self.statement())

        if result.error:
            return result

        statements.append(statement)

        more_statements = True

        while True:
            newline_count = 0

            while self.current_token.type == TT_NEWLINE:
                result.registerAdvancement()
                self.advance()

                newline_count += 1

            if newline_count == 0:
                more_statements = False

            if not more_statements:
                break

            statement = result.tryRegister(self.statement())

            if not statement:
                self.reverse(result.to_reverse_count)
                more_statements = False

                continue

            statements.append(statement)

        return result.success(ListNode(
            statements,
            pos_start,
            self.current_token.pos_end.copy()
        ))

    def call(self):
        result = ParseResult()
        atom = result.register(self.atom())

        if result.error:
            return result

        if self.current_token.type == TT_LPAREN:
            result.registerAdvancement()
            self.advance()

            arg_nodes = []

            if self.current_token.type == TT_RPAREN:
                result.registerAdvancement()
                self.advance()

            else:
                arg_nodes.append(result.register(self.expr()))

                if result.error:
                    return result.failure(InvalidSyntaxError(
                        self.current_token.pos_start, self.current_token.pos_end,
                        'Expected ")", "object", "if", "walk", "while", "func", int, float, identifier, "+", "-", '
                        '"(", "[" or "oppositeof"'
                    ))

                while self.current_token.type == TT_COMMA:
                    result.registerAdvancement()
                    self.advance()

                    arg_nodes.append(result.register(self.expr()))

                    if result.error:
                        return result

                if self.current_token.type != TT_RPAREN:
                    return result.failure(InvalidSyntaxError(
                        self.current_token.pos_start, self.current_token.pos_end,
                        'Expected "," or ")"'
                    ))

                result.registerAdvancement()
                self.advance()

            return result.success(CallNode(atom, arg_nodes))

        return result.success(atom)

    def atom(self):
        result = ParseResult()
        token = self.current_token

        if token.type in (TT_INT, TT_FLOAT):
            result.registerAdvancement()
            self.advance()
            return result.success(NumberNode(token))

        if token.type == TT_STR:
            result.registerAdvancement()
            self.advance()

            return result.success(StringNode(token))

        elif token.type == TT_IDENTIFIER:
            result.registerAdvancement()
            self.advance()
            return result.success(VariableAccessNode(token))

        elif token.type == TT_LPAREN:
            result.registerAdvancement()
            self.advance()
            expr = result.register(self.expr())

            if result.error:
                return result

            if self.current_token.type == TT_RPAREN:
                result.registerAdvancement()
                self.advance()
                return result.success(expr)

            else:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected ")"'
                ))

        elif token.type == TT_LSQUARE:
            list_expr = result.register(self.listExpr())

            if result.error:
                return result

            return result.success(list_expr)

        elif token.matches(TT_KEYWORD, 'if'):
            if_expr = result.register(self.ifExpr())

            if result.error:
                return result

            return result.success(if_expr)

        elif token.matches(TT_KEYWORD, 'walk'):
            for_expr = result.register(self.forExpr())

            if result.error:
                return result

            return result.success(for_expr)

        elif token.matches(TT_KEYWORD, 'while'):
            while_expr = result.register(self.whileExpr())

            if result.error:
                return result

            return result.success(while_expr)

        elif token.matches(TT_KEYWORD, 'func'):
            func_def = result.register(self.functionDefinition())

            if result.error:
                return result

            return result.success(func_def)

        return result.failure(InvalidSyntaxError(
            token.pos_start, token.pos_end, 'Expected "object", "if", "walk", "while", "func", int, float, identifier, '
                                            '"+", "-" "(" or "["'
        ))

    def power(self):
        return self.binaryOperator(self.call, (TT_POW,), self.factor)

    def factor(self) -> ParseResult:
        result = ParseResult()
        token = self.current_token

        if token.type in (TT_PLUS, TT_MINUS):
            result.registerAdvancement()
            self.advance()
            factor = result.register(self.factor())

            if result.error:
                return result

            return result.success(UnaryOperatorNode(token, factor))

        return self.power()

    def term(self):
        return self.binaryOperator(self.factor, (TT_MUL, TT_DIV))

    def functionDefinition(self):
        result = ParseResult()

        if not self.current_token.matches(TT_KEYWORD, 'func'):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "func"'
            ))

        result.registerAdvancement()
        self.advance()

        if self.current_token.type == TT_IDENTIFIER:
            var_name_token = self.current_token
            result.registerAdvancement()
            self.advance()

            if self.current_token.type != TT_LPAREN:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected "("'
                ))

        else:
            var_name_token = None
            if self.current_token.type != TT_LPAREN:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected identifier or "("'
                ))

        result.registerAdvancement()
        self.advance()

        arg_name_tokens = []

        if self.current_token.type == TT_IDENTIFIER:
            arg_name_tokens.append(self.current_token)

            result.registerAdvancement()
            self.advance()

            while self.current_token.type == TT_COMMA:
                result.registerAdvancement()
                self.advance()

                if self.current_token.type != TT_IDENTIFIER:
                    return result.failure(InvalidSyntaxError(
                        self.current_token.pos_start, self.current_token.pos_end,
                        'Expected Identifier'
                    ))

                arg_name_tokens.append(self.current_token)

                result.registerAdvancement()
                self.advance()

            if self.current_token.type != TT_RPAREN:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected "," or ")"'
                ))

        else:
            if self.current_token.type != TT_RPAREN:
                return result.failure(InvalidSyntaxError(
                    self.current_token.pos_start, self.current_token.pos_end,
                    'Expected identifier or ")"'
                ))

        result.registerAdvancement()
        self.advance()

        if self.current_token.type == TT_ARROW:
            result.registerAdvancement()
            self.advance()

            node_to_return = result.register(self.expr())

            if result.error:
                return result

            return result.success(FunctionDefinitionNode(
                var_name_token,
                arg_name_tokens,
                node_to_return,
                True
            ))

        if self.current_token.type != TT_NEWLINE:
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "->" or newline'
            ))

        result.registerAdvancement()
        self.advance()

        body = result.register(self.statements())

        if result.error:
            return result

        if not self.current_token.matches(TT_KEYWORD, 'endbody'):
            return result.failure(InvalidSyntaxError(
                self.current_token.pos_start, self.current_token.pos_end,
                'Expected "endbody"'
            ))

        result.registerAdvancement()
        self.advance()

        return result.success(FunctionDefinitionNode(
            var_name_token,
            arg_name_tokens,
            body,
            False
        ))

    def binaryOperator(self, func_a, ops, func_b=None):
        if func_b is None:
            func_b = func_a

        result = ParseResult()
        left = result.register(func_a())
        if result.error:
            return result

        while self.current_token.type in ops or (self.current_token.type, self.current_token.value) in ops:
            op_token = self.current_token
            result.registerAdvancement()
            self.advance()
            right = result.register(func_b())

            if result.error:
                return result

            left = BinaryOperatorNode(left, op_token, right)

        return result.success(left)

File: src/parsing/parser.rs
use crate::{
    errors::standard_error::StandardError,
    lexing::{position::Position, token::Token, token_type::TokenType},
    nodes::{
        binary_operator_node::BinaryOperatorNode, common_node::CommonNode, number_node::NumberNode,
        return_node::ReturnNode, string_node::StringNode, unary_operator_node::UnaryOperatorNode,
        variable_assign_node::VariableAssignNode,
    },
    parsing::parse_result::ParseResult,
};

pub struct Parser {
    pub tokens: Vec<Token>,
    pub token_index: isize,
    pub current_token: Option<Token>,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        let mut parser = Parser {
            tokens: tokens,
            token_index: -1,
            current_token: None,
        };
        parser.advance();

        parser
    }

    pub fn advance(&mut self) -> Option<Token> {
        self.token_index += 1;
        self.update_current_token();

        self.current_token.clone()
    }

    pub fn reverse(&mut self, amount: usize) -> Option<Token> {
        self.token_index -= amount as isize;
        self.update_current_token();

        self.current_token.clone()
    }

    pub fn update_current_token(&mut self) {
        if self.token_index >= 0 && self.token_index < self.tokens.len() as isize {
            self.current_token = Some(self.tokens[self.token_index as usize].clone());
        }
    }

    pub fn current_pos_start(&self) -> Position {
        self.current_token
            .as_ref()
            .and_then(|tok| tok.pos_start.as_ref())
            .cloned()
            .expect("Expected a pos_start")
    }

    pub fn current_pos_end(&self) -> Position {
        self.current_token
            .as_ref()
            .and_then(|tok| tok.pos_end.as_ref())
            .cloned()
            .expect("Expected a pos_end")
    }

    pub fn parse(&mut self) -> ParseResult {
        let mut parse_result = self.statements();

        if parse_result.error.is_some()
            && self.current_token.as_ref().unwrap().token_type != TokenType::TT_EOF
        {
            return parse_result.failure(Some(StandardError::new(
                "expected operator or bracket".to_string(),
                self.current_pos_start(),
                self.current_pos_end(),
                Some("add one of the following: '+', '-', '*', '/', or '}'".to_string()),
            )));
        }

        parse_result
    }

    pub fn comparison_expr(&mut self) -> ParseResult {
        let mut parse_result = ParseResult::new();

        if self
            .current_token
            .as_ref()
            .unwrap()
            .matches(TokenType::TT_KEYWORD, Some("oppositeof"))
        {
            let op_token = self.current_token.as_ref().unwrap().clone();
            parse_result.register_advancement();
            self.advance();

            let node = parse_result
                .register(self.comparison_expr())
                .unwrap()
                .clone();

            if parse_result.error.is_some() {
                return parse_result;
            }

            return parse_result.success(Some(Box::new(UnaryOperatorNode::new(
                op_token.clone(),
                node.clone(),
            )) as Box<dyn CommonNode>));
        }

        let node = parse_result.register(self.binary_operator(
            "arithmetic_expr",
            vec![
                (TokenType::TT_EE, ""),
                (TokenType::TT_NE, ""),
                (TokenType::TT_LT, ""),
                (TokenType::TT_GT, ""),
                (TokenType::TT_LTE, ""),
                (TokenType::TT_GTE, ""),
            ],
            None,
        ));

        if parse_result.error.is_some() {
            return parse_result.failure(Some(StandardError::new(
                "expected an object or operator".to_string(),
                self.current_pos_start(),
                self.current_pos_end(),
                Some("add one of the following: integer, float, identifier, 'oppositeof', '+', '-', '(', or '['".to_string()),
            )));
        }

        parse_result.success(node)
    }

    pub fn arithmetic_expr(&mut self) -> ParseResult {
        self.binary_operator(
            "term",
            vec![(TokenType::TT_PLUS, ""), (TokenType::TT_MINUS, "")],
            None,
        )
    }

    pub fn expr(&mut self) -> ParseResult {
        let mut parse_result = ParseResult::new();

        if self
            .current_token
            .as_ref()
            .unwrap()
            .matches(TokenType::TT_KEYWORD, Some("obj"))
        {
            parse_result.register_advancement();
            self.advance();

            if self.current_token.as_ref().unwrap().token_type != TokenType::TT_IDENTIFIER {
                return parse_result.failure(Some(StandardError::new(
                    "expected identifier".to_string(),
                    self.current_pos_start(),
                    self.current_pos_end(),
                    Some("add a name for this object like 'hotdog'".to_string()),
                )));
            }

            let var_name = self.current_token.clone().unwrap();
            parse_result.register_advancement();
            self.advance();

            if self.current_token.as_ref().unwrap().token_type != TokenType::TT_EQ {
                return parse_result.failure(Some(StandardError::new(
                    "expected '='".to_string(),
                    self.current_pos_start(),
                    self.current_pos_end(),
                    Some(
                        format!(
                            "add a '=' to set the value of the variable '{}'",
                            var_name.value.unwrap().clone()
                        )
                        .to_string(),
                    ),
                )));
            }

            parse_result.register_advancement();
            self.advance();
            let expr = parse_result.register(self.expr());

            if parse_result.error.is_some() {
                return parse_result;
            }

            return parse_result.success(Some(Box::new(VariableAssignNode::new(
                var_name.clone(),
                expr.unwrap(),
            )) as Box<dyn CommonNode>));
        }

        let node = parse_result.register(self.binary_operator(
            "comparison_expr",
            vec![
                (TokenType::TT_KEYWORD, "and"),
                (TokenType::TT_KEYWORD, "or"),
            ],
            None,
        ));

        if parse_result.error.is_some() {
            return parse_result.failure(Some(StandardError::new(
                "expected a keyword, type, or function".to_string(),
                self.current_pos_start(),
                self.current_pos_end(),
                Some("add one of the following: 'obj', 'if', 'walk', 'while', 'func', 'oppositeof', integer, float, identifier, '+', '-', '(', or '['".to_string()),
            )));
        }

        parse_result.success(node)
    }

    pub fn statement(&mut self) -> ParseResult {
        let mut parse_result = ParseResult::new();
        let pos_start = self.current_pos_start();

        if self
            .current_token
            .as_ref()
            .unwrap()
            .matches(TokenType::TT_KEYWORD, Some("give"))
        {
            parse_result.register_advancement();
            self.advance();

            let expr = parse_result.try_register(self.expr());

            if expr.is_none() {
                self.reverse(parse_result.to_reverse_count);
            }

            return parse_result.success(Some(Box::new(ReturnNode::new(
                expr.unwrap(),
                Some(pos_start),
                Some(self.current_pos_start()),
            )) as Box<dyn CommonNode>));
        }

        let expr = parse_result.register(self.expr());

        if parse_result.error.is_some() {
            return parse_result.failure(Some(StandardError::new(
                "expected keyword, object, or operator".to_string(),
                self.current_pos_start(),
                self.current_pos_end(),
                Some("add any of the following: 'give', 'next', 'leave', 'obj', 'oppositeof', 'if', 'walk', 'while', 'func', int, float, identifier, '+', '-', '(', or '['".to_string()),
            )));
        }

        parse_result.success(expr)
    }

    pub fn statements(&mut self) -> ParseResult {
        let mut parse_result = ParseResult::new();
        let statements: Vec<ParseResult> = Vec::new();
        let pos_start = self.current_pos_start();

        while self.current_token.as_mut().unwrap().token_type == TokenType::TT_NEWLINE {
            parse_result.register_advancement();
            self.advance();
        }

        let statement = parse_result.register(self.statement());

        parse_result
    }

    pub fn call(&mut self) -> ParseResult {
        let mut parse_result = ParseResult::new();
        let atom = parse_result.register(self.atom());

        if parse_result.error.is_some() {
            return parse_result;
        }

        parse_result.success(atom)
    }

    pub fn atom(&mut self) -> ParseResult {
        let mut parse_result = ParseResult::new();
        let token = self.current_token.as_ref().unwrap().clone();

        if [TokenType::TT_INT, TokenType::TT_FLOAT].contains(&token.token_type) {
            parse_result.register_advancement();
            self.advance();

            return parse_result
                .success(Some(Box::new(NumberNode::new(token)) as Box<dyn CommonNode>));
        } else if token.token_type == TokenType::TT_STR {
            parse_result.register_advancement();
            self.advance();

            return parse_result
                .success(Some(Box::new(StringNode::new(token)) as Box<dyn CommonNode>));
        }

        parse_result.failure(Some(StandardError::new(
            "expected object, keyword, function, or type".to_string(),
            token.pos_start.unwrap(),
            token.pos_end.unwrap(),
            Some("add any of the following: 'obj', 'if', 'walk', 'while', 'func', integer, float, identifier, '+' , '-' , '(' or '['".to_string()),
        )))
    }

    pub fn power(&mut self) -> ParseResult {
        self.binary_operator("call", vec![(TokenType::TT_POW, "")], Some("factor"))
    }

    pub fn factor(&mut self) -> ParseResult {
        let mut parse_result = ParseResult::new();
        let token = self.current_token.as_ref().unwrap().clone();

        if [TokenType::TT_PLUS, TokenType::TT_MINUS].contains(&token.token_type) {
            parse_result.register_advancement();
            self.advance();
            let factor = parse_result.register(self.factor());

            if parse_result.error.is_some() {
                return parse_result;
            }

            return parse_result.success(Some(Box::new(UnaryOperatorNode::new(
                token,
                factor.unwrap(),
            ))));
        }

        return self.power();
    }

    pub fn term(&mut self) -> ParseResult {
        self.binary_operator(
            "factor",
            vec![(TokenType::TT_MUL, ""), (TokenType::TT_DIV, "")],
            None,
        )
    }

    pub fn binary_operator(
        &mut self,
        func_a: &'static str,
        ops: Vec<(TokenType, &'static str)>,
        func_b: Option<&'static str>,
    ) -> ParseResult {
        let func_b = func_b.unwrap_or_else(|| func_a);

        let mut parse_result = ParseResult::new();
        let mut left = parse_result.register(match func_a {
            "comparison_expr" => self.comparison_expr(),
            "arithmetic_expr" => self.arithmetic_expr(),
            "term" => self.term(),
            "factor" => self.factor(),
            "call" => self.call(),
            _ => panic!("CRITICAL ERROR: GLANG COULD NOT FIND EXPRESSION IN BINARY OPERATOR"),
        });

        if parse_result.error.is_some() {
            return parse_result;
        }

        while ops.contains(&(
            self.current_token.clone().unwrap().token_type,
            self.current_token
                .clone()
                .unwrap()
                .value
                .unwrap_or_else(|| "".to_string())
                .as_str(),
        )) || ops.contains(&(self.current_token.clone().unwrap().token_type, ""))
        {
            let op_token = self.current_token.clone().unwrap().clone();
            parse_result.register_advancement();
            self.advance();
            let right = parse_result.register(match func_b {
                "comparison_expr" => self.comparison_expr(),
                "arithmetic_expr" => self.arithmetic_expr(),
                "term" => self.term(),
                "factor" => self.factor(),
                "call" => self.call(),
                _ => panic!("CRITICAL ERROR: GLANG COULD NOT FIND EXPRESSION IN BINARY OPERATOR"),
            });

            if parse_result.error.is_some() {
                return parse_result;
            }

            left = Some(Box::new(BinaryOperatorNode::new(
                left.unwrap().clone(),
                op_token,
                right.unwrap(),
            )) as Box<dyn CommonNode>);
        }

        parse_result.success(left)
    }
}
